Handle broader ConnectException than HttpHostConnectException for HTTP failover

From https://github.com/GoogleContainerTools/jib/issues/1512#issuecomment-468078006

> the JIB v1.0.0 can handle the HTTP and HTTPS connection, but v1.0.1 can handle the HTTPS connection only (it only tries at HTTPS)

Between 1.0.0 and 1.0.1, there has been no code change on the Jib side regarding handling this, but we did upgrade the Google HTTP Client: https://github.com/GoogleContainerTools/jib/commit/f2fe6aee6a32d0e23ab2de62801d95b10c86a103

And as can be seen in the above commit, the upgrade has surfaced the behavioral changes of Apache HttpClient (throwing a more general `SSLException` than `SSLPeerUnverifiedException`). From the log in https://github.com/GoogleContainerTools/jib/issues/1512#issuecomment-468078006, the new Apache HttpClient seems to have brought a similar behavioral change in this case too:

```
Caused by: java.net.ConnectException: Connection refused (Connection refused)
    at java.net.PlainSocketImpl.socketConnect (Native Method)
    at java.net.AbstractPlainSocketImpl.doConnect (AbstractPlainSocketImpl.java:399)
    at java.net.AbstractPlainSocketImpl.connectToAddress (AbstractPlainSocketImpl.java:242)
    at java.net.AbstractPlainSocketImpl.connect (AbstractPlainSocketImpl.java:224)
    at java.net.SocksSocketImpl.connect (SocksSocketImpl.java:403)
    at java.net.Socket.connect (Socket.java:591)
    at org.apache.http.conn.ssl.SSLSocketFactory.connectSocket (SSLSocketFactory.java:542)
```

`ConnectException` is a super-class of `HttpHostConnectException`. I think we should catch the former now.

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
@@ -31,6 +31,7 @@ import com.google.cloud.tools.jib.registry.json.ErrorEntryTemplate;
 import com.google.cloud.tools.jib.registry.json.ErrorResponseTemplate;
 import com.google.common.annotations.VisibleForTesting;
 import java.io.IOException;
+import java.net.ConnectException;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.security.GeneralSecurityException;
@@ -38,7 +39,6 @@ import java.util.function.Function;
 import javax.annotation.Nullable;
 import javax.net.ssl.SSLException;
 import org.apache.http.NoHttpResponseException;
-import org.apache.http.conn.HttpHostConnectException;
 
 /**
  * Makes requests to a registry endpoint.
@@ -154,7 +154,7 @@ class RegistryEndpointCaller<T> {
     } catch (SSLException ex) {
       return handleUnverifiableServerException(url);
 
-    } catch (HttpHostConnectException ex) {
+    } catch (ConnectException ex) {
       if (allowInsecureRegistries && isHttpsProtocol(url) && url.getPort() == -1) {
         // Fall back to HTTP only if "url" had no port specified (i.e., we tried the default HTTPS
         // port 443) and we could not connect to 443. It's worth trying port 80.

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
@@ -136,7 +136,6 @@ public class RegistryEndpointCallerTest {
 
     try {
       secureEndpointCaller.call();
-
       Assert.fail("Secure caller should fail if cannot verify server");
     } catch (InsecureRegistryException ex) {
       Assert.assertEquals(

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
@@ -30,6 +30,7 @@ import com.google.cloud.tools.jib.http.Connection;
 import com.google.cloud.tools.jib.http.MockConnection;
 import com.google.cloud.tools.jib.http.Response;
 import java.io.IOException;
+import java.net.ConnectException;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.util.Collections;
@@ -38,7 +39,6 @@ import java.util.function.Function;
 import javax.annotation.Nullable;
 import javax.net.ssl.SSLPeerUnverifiedException;
 import org.apache.http.NoHttpResponseException;
-import org.apache.http.conn.HttpHostConnectException;
 import org.hamcrest.CoreMatchers;
 import org.junit.After;
 import org.junit.Assert;
@@ -136,6 +136,7 @@ public class RegistryEndpointCallerTest {
 
     try {
       secureEndpointCaller.call();
+
       Assert.fail("Secure caller should fail if cannot verify server");
     } catch (InsecureRegistryException ex) {
       Assert.assertEquals(
@@ -206,7 +207,7 @@ public class RegistryEndpointCallerTest {
   public void testCall_insecureCallerOnHttpServerAndNoPortSpecified()
       throws IOException, RegistryException {
     Mockito.when(mockConnection.send(Mockito.eq("httpMethod"), Mockito.any()))
-        .thenThrow(Mockito.mock(HttpHostConnectException.class)) // server is not listening on 443
+        .thenThrow(Mockito.mock(ConnectException.class)) // server is not listening on 443
         .thenReturn(mockResponse); // respond when connected through 80
 
     RegistryEndpointCaller<String> insecureEndpointCaller = createRegistryEndpointCaller(true, -1);
@@ -230,12 +231,12 @@ public class RegistryEndpointCallerTest {
   public void testCall_secureCallerOnNonListeningServerAndNoPortSpecified()
       throws IOException, RegistryException {
     Mockito.when(mockConnection.send(Mockito.eq("httpMethod"), Mockito.any()))
-        .thenThrow(Mockito.mock(HttpHostConnectException.class)); // server is not listening on 443
+        .thenThrow(Mockito.mock(ConnectException.class)); // server is not listening on 443
 
     try {
       secureEndpointCaller.call();
       Assert.fail("Should not fall back to HTTP if not allowInsecureRegistries");
-    } catch (HttpHostConnectException ex) {
+    } catch (ConnectException ex) {
       Assert.assertNull(ex.getMessage());
     }
 
@@ -251,14 +252,14 @@ public class RegistryEndpointCallerTest {
   public void testCall_insecureCallerOnNonListeningServerAndPortSpecified()
       throws IOException, RegistryException {
     Mockito.when(mockConnection.send(Mockito.eq("httpMethod"), Mockito.any()))
-        .thenThrow(Mockito.mock(HttpHostConnectException.class)); // server is not listening on 5000
+        .thenThrow(Mockito.mock(ConnectException.class)); // server is not listening on 5000
 
     RegistryEndpointCaller<String> insecureEndpointCaller =
         createRegistryEndpointCaller(true, 5000);
     try {
       insecureEndpointCaller.call();
       Assert.fail("Should not fall back to HTTP if port was explicitly given and cannot connect");
-    } catch (HttpHostConnectException ex) {
+    } catch (ConnectException ex) {
       Assert.assertNull(ex.getMessage());
     }
 

--- a/jib-gradle-plugin/CHANGELOG.md
+++ b/jib-gradle-plugin/CHANGELOG.md
@@ -11,6 +11,8 @@ All notable changes to this project will be documented in this file.
 
 ### Fixed
 
+- Fixed an issue where setting `allowInsecureRegistries` may fail to try HTTP.
+
 ## 1.0.1
 
 ### Added

--- a/jib-maven-plugin/CHANGELOG.md
+++ b/jib-maven-plugin/CHANGELOG.md
@@ -11,6 +11,8 @@ All notable changes to this project will be documented in this file.
 
 ### Fixed
 
+- Fixed an issue where setting `<allowInsecureRegistries>` may fail to try HTTP.
+
 ## 1.0.1
 
 ### Added

